src/main/java/org/spongepowered/common/data/processor/data/tileentity/FurnaceDataProcessor.java

```diff
@@ -0,0 +1,142 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.processor.data.tileentity;
+
+import net.minecraft.tileentity.TileEntityFurnace;
+
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataTransactionBuilder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.tileentity.ImmutableFurnaceData;
+import org.spongepowered.api.data.manipulator.mutable.tileentity.FurnaceData;
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.common.data.manipulator.immutable.tileentity.ImmutableSpongeFurnaceData;
+import org.spongepowered.common.data.manipulator.mutable.tileentity.SpongeFurnaceData;
+import org.spongepowered.common.data.processor.common.AbstractSpongeDataProcessor;
+import org.spongepowered.common.data.util.DataUtil;
+
+import java.util.Optional;
+
+public class FurnaceDataProcessor extends AbstractSpongeDataProcessor<FurnaceData, ImmutableFurnaceData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof TileEntityFurnace;
+    }
+
+    @Override
+    public Optional<FurnaceData> from(DataHolder dataHolder) {
+        if (this.supports(dataHolder)) {
+            final TileEntityFurnace tileEntityFurnace = (TileEntityFurnace) dataHolder;
+            final int remainingBurnTime = tileEntityFurnace.getField(0);
+            final int remainingCookTime = tileEntityFurnace.getField(3) - tileEntityFurnace.getField(2);
+            final int maxCookTime = tileEntityFurnace.getField(3);
+            return Optional.of(new SpongeFurnaceData(remainingBurnTime, remainingCookTime, maxCookTime));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<FurnaceData> createFrom(DataHolder dataHolder) {
+        if (this.supports(dataHolder)) {
+            return this.from(dataHolder);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<FurnaceData> fill(DataHolder dataHolder, FurnaceData manipulator, MergeFunction overlap) {
+        if (this.supports(dataHolder)) {
+            final Optional<FurnaceData> dataOptional = this.from(dataHolder);
+            if (dataOptional.isPresent()) {
+                return Optional.of(overlap.merge(manipulator, dataOptional.get()));
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<FurnaceData> fill(DataContainer container, FurnaceData furnaceData) {
+        DataUtil.checkDataExists(container, Keys.REMAINING_BURN_TIME.getQuery());
```

Comment:
You should check if these exist manually, and return `Optional.absent` if they don't.